src/libaktualizr/crypto/crypto.cc

```diff
@@ -345,3 +397,37 @@ bool Crypto::generateKeyPair(KeyType key_type, std::string *public_key, std::str
   }
   return Crypto::generateRSAKeyPair(key_type, public_key, private_key);
 }
+
+bool Crypto::IsRsaKeyType(KeyType type) {
+  switch (type) {
+    case kRSA2048:
+    case kRSA4096:
+      return true;
+    default:
+      return false;
+  }
+}
+KeyType Crypto::IdentifyRSAKeyType(const std::string &public_key_pem) {
+  StructGuard<BIO> bufio(BIO_new_mem_buf(reinterpret_cast<const void *>(public_key_pem.c_str()),
+                                         static_cast<int>(public_key_pem.length())),
+                         BIO_vfree);
+  if (bufio.get() == nullptr) {
+    throw std::runtime_error("BIO_new_mem_buf failed");
+  }
+  StructGuard<::RSA> rsa(PEM_read_bio_RSA_PUBKEY(bufio.get(), nullptr, nullptr, nullptr), RSA_free);
+
+  if (rsa.get() == nullptr) {
+    return kUnknownKey;
+  }
+
+  int key_length = RSA_size(rsa.get());
+
+  // It is not clear from the OpenSSL documentation if RSA_size returns
+  // exactly 2048 or 4096, or if this can vary. Instead of an exact match,
+  // just split the range into two
+  if (key_length < 3000) {
+    return kRSA2048;
+  } else {
+    return kRSA4096;
+  }
```

Comment:
This seems dangerous to me. From what I can tell, RSA_size returns whatever the modulus is, which should be equivalent to our expectations (exactly 2048 or 4096). Since key sizes could potentially be whatever you can imagine, this seems rife for error.

Reply:
Reasonable. I'll check for an absolute match.